From 742f715bb28d11d44e77e8b8a21a42f696957acd Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Rapha=C3=ABl=20M=C3=A9lotte?= <raphael.melotte@mind.be>
Date: Wed, 31 Aug 2022 16:48:37 +0200
Subject: [PATCH 6/6] iw: add station flags
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

This can be used for example to manually set a station as authorized.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>
---
 station.c | 39 +++++++++++++++++++++++++++++++++++++++
 1 file changed, 39 insertions(+)

diff --git a/station.c b/station.c
index 6a12ae0..4076c88 100644
--- a/station.c
+++ b/station.c
@@ -900,3 +900,42 @@ static int handle_station_new(struct nl80211_state *state,
 COMMAND(station, new, "<MAC> <listen_interval> <supported_rates> <aid> <flags>",
 	NL80211_CMD_NEW_STATION, 0, CIB_NETDEV, handle_station_new,
 	"Add a new station\n.");
+
+static int handle_station_flags(struct nl80211_state *state,
+			       struct nl_msg *msg,
+			       int argc, char **argv,
+			       enum id_input id)
+{
+	char *end;
+	unsigned char mac[6];
+	struct nl_msg *key_msg;
+	int err = -EINVAL;
+
+	key_msg = nlmsg_alloc();
+	if (!key_msg)
+		return -ENOMEM;
+
+	/* mac */
+	if (argc) {
+		if (mac_addr_a2n(mac, argv[0]) == 0) {
+			NLA_PUT(msg, NL80211_ATTR_MAC, 6, mac);
+			argv++;
+			argc--;
+		}
+	}
+
+	/* flags2 */
+	if (argc) {
+		nla_put_u64(msg, NL80211_ATTR_STA_FLAGS2, strtoull(argv[0], &end, 10));
+		argv++;
+		argc--;
+	}
+	err = 0;
+
+ nla_put_failure:
+	nlmsg_free(key_msg);
+	return err;
+}
+COMMAND(station, set_flags, "<MAC> <flags>",
+	NL80211_CMD_SET_STATION, 0, CIB_NETDEV, handle_station_flags,
+	"Set station flags\n.");
-- 
2.37.1

